keyring/backends/Windows/__init__.py

```diff
@@ -109,26 +101,32 @@ def _get_password(self, target):
             res = win32cred.CredRead(
                 Type=win32cred.CRED_TYPE_GENERIC, TargetName=target
             )
-        except pywintypes.error as e:
-            if e.winerror == 1168 and e.funcname == 'CredRead':  # not found
+        except OSError as e:
+            if e.winerror == 1168:  # not found
                 return None
             raise
+
         return DecodingCredential(res)
 
-    def set_password(self, service, username, password):
+    def set_password(self, service, username, password, encoding='utf-16-le'):
         existing_pw = self._get_password(service)
         if existing_pw:
-            # resave the existing password using a compound target
             existing_username = existing_pw['UserName']
-            target = self._compound_name(existing_username, service)
-            self._set_password(
-                target,
-                existing_username,
-                existing_pw.value,
-            )
-        self._set_password(service, username, str(password))
-
-    def _set_password(self, target, username, password):
+            # resave the existing password using a compound target
+            # Fixes part of https://github.com/jaraco/keyring/issues/545,
+            # but get_credentials also needs to be fixed to search in the same
+            # order as get_password.
+            if existing_username != username:
+                target = self._compound_name(existing_username, service)
+                self._set_password(
+                    target,
+                    existing_username,
+                    existing_pw.value,
+                    encoding=encoding,
+                )
+        self._set_password(service, username, str(password), encoding=encoding)
```

Comment:
I'm confused by this comment. It's saying that by re-saving the existing password using a compound target, it fixes part of 545, yet re-saving the existing password was already part of the prior change. Moreover, "get_credentials also needs to be fixed" feels like a misplaced TODO. Additionally, the comment doesn't say anything about what the code is meant to do, only has a reference to a reported issue.

By reading the issue, I now think the comment is only about the "if existing_username" check. The problem with putting the comment here is it becomes interleaved with the other comment, making it difficult to tell what comment goes with which parts of code. Since this change adds a whole new branch of code, it may be time to consider a refactor.

At the very least, I'd move the "resave" comment back to where it was (at the beginning of the `if existing_pw` branch and then move the rest of the comment into the `if existing_username` branch and rephrase to indicate its essential purpose, something like, "Only set the compound target when usernames match. Ref #545."

Reply:
Agreed.